src/frontend/src/optimizer/plan_node/logical_changed_log.rs

```diff
@@ -0,0 +1,150 @@
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalChangedLog {
+    pub base: PlanBase<Logical>,
+    core: generic::ChangedLog<PlanRef>,
+}
+
+impl LogicalChangedLog {
+    pub fn create(input: PlanRef) -> PlanRef {
+        Self::new(input, true, true).into()
+    }
+
+    pub fn new(input: PlanRef, need_op: bool, need_changed_log_row_id: bool) -> Self {
+        let core = generic::ChangedLog::new(input, need_op, need_changed_log_row_id);
+        Self::with_core(core)
+    }
+
+    pub fn with_core(core: generic::ChangedLog<PlanRef>) -> Self {
+        let base = PlanBase::new_logical_with_core(&core);
+        LogicalChangedLog { base, core }
+    }
+}
+
+impl PlanTreeNodeUnary for LogicalChangedLog {
+    fn input(&self) -> PlanRef {
+        self.core.input.clone()
+    }
+
+    fn clone_with_input(&self, input: PlanRef) -> Self {
+        Self::new(input, self.core.need_op, self.core.need_changed_log_row_id)
+    }
+
+    fn rewrite_with_input(
+        &self,
+        input: PlanRef,
+        input_col_change: ColIndexMapping,
+    ) -> (Self, ColIndexMapping) {
+        let changed_log = Self::new(input, self.core.need_op, true);
+        if self.core.need_op {
+            let mut output_vec = input_col_change.to_parts().0.to_vec();
+            let len = input_col_change.to_parts().1;
+            output_vec.push(Some(len));
+            let out_col_change = ColIndexMapping::new(output_vec, len + 1);
+            (changed_log, out_col_change)
+        } else {
+            (changed_log, input_col_change)
+        }
+    }
+}
+
+impl_plan_tree_node_for_unary! {LogicalChangedLog}
+impl_distill_by_unit!(LogicalChangedLog, core, "LogicalChangedLog");
+
+impl ExprRewritable for LogicalChangedLog {}
+
+impl ExprVisitable for LogicalChangedLog {}
+
+impl PredicatePushdown for LogicalChangedLog {
+    fn predicate_pushdown(
+        &self,
+        predicate: Condition,
+        ctx: &mut super::PredicatePushdownContext,
+    ) -> PlanRef {
+        gen_filter_and_pushdown(self, predicate, Condition::true_cond(), ctx)
+    }
+}
+
+impl ColPrunable for LogicalChangedLog {
+    fn prune_col(&self, required_cols: &[usize], ctx: &mut ColumnPruningContext) -> PlanRef {
+        let mut new_required_cols = required_cols.to_vec();
+        let fields = self.schema().fields();
+        let mut need_op = false;
+        let mut need_changed_log_row_id = false;
+        required_cols.iter().for_each(|a| {
+            if let Some(f) = fields.get(*a) {
+                if f.name == "op" {
+                    new_required_cols.pop();
+                    need_op = true;
+                } else if f.name == "_changed_log_row_id" {
+                    new_required_cols.pop();
+                    need_changed_log_row_id = true;
+                }
+            }
+        });
+
+        let new_input = self.input().prune_col(&new_required_cols, ctx);
+        Self::new(new_input, need_op, need_changed_log_row_id).into()
+    }
```

Comment:
If `prune_col` could set `need_op` and `need_changed_log_row_id` to false, it might cause changed_log has no stream_key after `logical_rewrite_for_stream`.